pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -46,3 +70,426 @@ func (s *discovery) Close() {}
+		Listeners: *listeners,
+	}
+	utils.WriteAsJSON(w, resp)
+}
+
+var allowedEnvironmentVariables = map[string]struct{}{
+	"PATH":                     {},
+	"PWD":                      {},
+	"GUNICORN_CMD_ARGS":        {},
+	"WSGI_APP":                 {},
+	"DD_SERVICE":               {},
+	"DD_TAGS":                  {},
+	"DD_INJECTION_ENABLED":     {},
+	"SPRING_APPLICATION_NAME":  {},
+	"SPRING_CONFIG_LOCATIONS":  {},
+	"SPRING_CONFIG_NAME":       {},
+	"SPRING_PROFILES_ACTIVE":   {},
+	"CORECLR_PROFILER_PATH":    {},
+	"CORECLR_ENABLE_PROFILING": {},
+	"JAVA_TOOL_OPTIONS":        {},
+	"_JAVA_OPTIONS":            {},
+	"JDK_JAVA_OPTIONS":         {},
+	"JAVA_OPTIONS":             {},
+	"CATALINA_OPTS":            {},
+	"JDPA_OPTS":                {},
+	"VIRTUAL_ENV":              {},
+}
+
+func filterEnv(in []string) (out []string) {
+	for _, env := range in {
+		split := strings.SplitN(env, "=", 2)
+		if len(split) != 2 {
+			continue
+		}
+
+		name := split[0]
+		if _, ok := allowedEnvironmentVariables[name]; ok {
+			out = append(out, env)
+		}
+	}
+	return
+}
+
+func (s *discovery) handleGetProc(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	pidStr := vars["pid"]
+	pid, err := strconv.ParseUint(pidStr, 10, 32)
+	if err != nil {
+		s.handleError(w, pathGetProc, http.StatusBadRequest, fmt.Errorf("failed to convert pid to integer: %v", err))
+		return
+	}
+
+	if _, err := os.Stat(path.Join(procfs.DefaultMountPoint, pidStr)); os.IsNotExist(err) {
+		s.handleError(w, pathGetProc, http.StatusNotFound, fmt.Errorf("/proc/{pid} does not exist: %v", err))
+		return
+	}
+	proc, err := procfs.NewProc(int(pid))
+	if err != nil {
+		s.handleError(w, pathGetProc, http.StatusInternalServerError, fmt.Errorf("failed to read procfs: %v", err))
+		return
+	}
+	env, err := proc.Environ()
+	if err != nil {
+		s.handleError(w, pathGetProc, http.StatusInternalServerError, fmt.Errorf("failed to read /proc/{pid}/environ: %v", err))
+		return
+	}
+	cwd, err := proc.Cwd()
+	if err != nil {
+		s.handleError(w, pathGetProc, http.StatusInternalServerError, fmt.Errorf("failed to read /proc/{pid}/cwd: %v", err))
+		return
+	}
+
+	resp := &model.GetProcResponse{
+		Proc: &model.Proc{
+			PID:     int(pid),
+			Environ: filterEnv(env),
+			CWD:     cwd,
+		},
+	}
+	utils.WriteAsJSON(w, resp)
+}
+
+// Lifted from
+// pkg/security/security_profile/activity_tree/process_node_snapshot.go.  The
+// socket inode information from /proc/../fd is needed to map the connection
+// from the net/tcp (and similar) files to actual ports.
+func getSockets(p *process.Process) ([]uint64, error) {
+	FDs, err := p.OpenFiles()
+	if err != nil {
+		return nil, err
+	}
+
+	// sockets have the following pattern "socket:[inode]"
+	var sockets []uint64
+	for _, fd := range FDs {
+		if strings.HasPrefix(fd.Path, "socket:[") {
+			sock, err := strconv.Atoi(strings.TrimPrefix(fd.Path[:len(fd.Path)-1], "socket:["))
+			if err != nil {
+				continue
+			}
+			if sock < 0 {
```

Comment:
The < check is gone after implementing Bryce's suggestion. I've not added the == 0 check, not sure it's really needed? The kernel doesn't use inode 0 for these socket inodes since some years ago but looks like an implementation detail: https://patchwork.kernel.org/project/linux-fsdevel/patch/1435245958-4507-1-git-send-email-cmaiolino@redhat.com/